### ISSUE TITLE
Help reproducing results for llama-2-7b. Low refusal rate 

### ISSUE THREAD

**Issue by bangawayoo** [GAIR-NLP/alignment-for-honesty#2]
Hi,
Thank you for your work.

I am trying to reproduce the results for LLaMA-2-7b for the Confidence-Verb configuration.
I am training on top of `meta-llama/Llama-2-7b-chat-hf` using the dataset provided in `data/training_data/triviaqa_7b.jsonl)`. Here are the steps I took:

1. Convert data using `train/process_training_data.py` with the following args:
  --sft_method SFTMethod.CONFIDENCE_VERB, 
  --prompt_id 3 
2. Train using fastchat using deepspeed. I am using the training setting providing in Section 5.3.
3.  Evaluate. Below are the results in TriviaQA 

|Model | Accuracy | Refusal Rate ('idk' outputs) | Prudence | Over-conserv.|
|---|---|---|---|---|
|Llama-2-7b-chat-hf| 69.3 | 0 | - | - |
|Confidence-verb (reproduced)| 71.1|5.2|7.2|4.9|
|Paper (Table 6) | 68.1 | NA | 56.0 | 11.4|

Since the evaluation code is implemented by me, there might be a slight difference. However, even after accounting for this, my fine-tuned version has a low refusal rate and higher task accuracy. Can you guess what I might be doing wrong? 

 

-----
Additionally, did you do any experiments using LoRA for aligning honesty? I am curious if LoRA would be capable of aligning honesty. Any help would be appreciated.

Thanks!

**Comment by ayyyq** [comment]
Thank you for your interest in our work. We provide our confidence-verb checkpoint (based on llama2-7b-chat) [here](https://huggingface.co/ayyyq/confucius-confidence-verb-7b) and its inference [here](https://drive.google.com/drive/folders/1YJi3ZOD-ZTnz8CR6yhpSTJmmlTLUu2_o?usp=sharing). You are welcome to use these resources to verify the accuracy of your inference and evaluation code. 

It seems that your training steps are correct so far. However, one thing you may need to pay attention to is **whether a space is needed after "Answer:" in the input**, as the handling of space may vary between Collie and other training frameworks. Additionally, **the prompts for the training and inference code need to match.**

If you have any further questions, please feel free to reach out to us. Thank you!

---
Unfortunately, we have not conducted experiments using LoRA. We would be very interested to hear if you or anyone has made progress in this application. Thank you for bringing this to our attention!

**Comment by bangawayoo** [comment]
Thank you for releasing these.
I will check them out to see whether my inference code is correct!

**Comment by bangawayoo** [comment]
@ayyyq 
I was able to reproduce the exact generation outputs with your model!

Seems like a training issue on my side. I have some questions that are not detailed in the paper. 
- Did you do left padding? or right padding during training? 
- Do you remember how many GPUs you used and the library for sharding, e.g. FSDP, Deepspeed?

**Comment by ayyyq** [comment]
Regarding your questions:
- We adopted the default left padding in [Collie v1.0.5](https://github.com/OpenMOSS/CoLLiE/tree/v1.0.5).
- Although I cannot confirm with absolute certainty, it‘s likely that we used 4 GPUs to train Llama2-7b-chat. It might be important to note that we used Collie's **Pipeline Parallelism** with a batch_size of 8. If you opt for Data Parallelism, you might consider setting dp_size * batch_size * gradient_accumulation_steps = 8.

**Comment by bangawayoo** [comment]
It seems like this was an issue related to the chat template. When using the original LLaMA-2 chat template, the model had a  low refusal rate. Training without the special tokens, e.g. [INST], [\INST], I was able to reproduce the numbers similar to those reported in the paper (refusal rate of ~30%).

Just in case anyone is interested, LLaMA-2-7b with LoRA also showed similar accuracy although its refusal rate was lower.
